sound/soc/intel/boards/sof_sdw_rt_amp.c

```diff
@@ -212,11 +213,25 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
 		}
 	}
 
-	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
-					  "%s spk:%s",
-					  card->components, codec_name);
-	if (!card->components)
-		return -ENOMEM;
+	/* Add components string for each codec dai */
+	for_each_rtd_codec_dais(rtd, i, dai) {
+		/*
+		 * rt_amp_spk_rtd_init() only for rt13xx amps. rt7xx multi-function codecs will
+		 * use their own rtd_init function.
+		 */
+		if (!strstr(dai->name, "rt13"))
+			continue;
+		snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai->name);
+		c_temp = devm_kasprintf(card->dev, GFP_KERNEL, "spk:%s", codec_name);
+		/* Check if the tag is already in card->components */
+		if (strstr(card->components, c_temp))
+			continue;
+		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
+						  "%s %s",
+						  card->components, c_temp);
+		if (!card->components)
```

Comment:
Lat time I have checked UCM only matches the first spk: ? or even it expects that it is is list of speaker codec names and not separate tags?

Reply:
> Lat time I have checked UCM only matches the first spk: ? or even it expects that it is is list of speaker codec names and not separate tags?

I haven't discussed with @shumingfan yet. But, I would expect they are separate tags. And of course we need UCM change, too.

Reply:
It should be a separate spk: + codec name pair. Currently only the first match is used (SpeakerCodec1), the second should be in SpeakerCodec2, third in SpeakerCodec3, etc.

Reply:
> It should be a separate spk: + codec name pair. Currently only the first match is used (SpeakerCodec1), the second should be in SpeakerCodec2, third in SpeakerCodec3, etc.

@shumingfan confirmed he can get SpeakerCodec2 and SpeakerCodec2, etc. So, "spk:rt712 spk:rt1320" should work.

Reply:
@bardliao @shumingfan How would userspace know which amplifier does what in a main/companion setup?